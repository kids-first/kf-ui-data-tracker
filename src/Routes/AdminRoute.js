import React from 'react';
import jwtDecode from 'jwt-decode';
import {Route, Redirect, withRouter} from 'react-router-dom';
import {Amplitude} from '@amplitude/react-amplitude';

export const isAdmin = () => {
  const token = localStorage.getItem('accessToken');
  if (token == null) {
    return false;
  }

  const decoded = jwtDecode(token);
  if (decoded.exp < Math.floor(new Date().getTime() / 1000)) return false;

  // Extract roles based on whether the token is from auth0
  const roles = decoded.context
    ? decoded.context.user.roles
    : decoded['https://kidsfirstdrc.org/roles'];

  return roles.includes('ADMIN');
};

/**
 * A route that only allows access to admin users, otherwise routes to the
 * index page
 */
const AdminRoute = ({component: Component, path, ...rest}) => {
  let comp = <Component />;

  return (
    <Amplitude
      eventProperties={{
        view: comp.type.WrappedComponent
          ? comp.type.WrappedComponent.name
          : comp.type.name,
        route: path,
      }}
    >
      <Route
        {...rest}
        path={path}
        render={props =>
          isAdmin() ? <Component {...props} /> : <Redirect to="/" />
        }
      />
    </Amplitude>
  );
};

export default withRouter(AdminRoute);
