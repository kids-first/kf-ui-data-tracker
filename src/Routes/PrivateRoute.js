import React from 'react';
import jwtDecode from 'jwt-decode';
import {Route, Redirect, withRouter} from 'react-router-dom';
import {Amplitude} from '@amplitude/react-amplitude';

export const hasToken = () => {
  const token = localStorage.getItem('accessToken');
  if (token == null) {
    return false;
  }
  return jwtDecode(token).exp > Math.floor(new Date().getTime() / 1000);
};

const PrivateRoute = ({component: Component, ...rest}) => {
  // note: unsure if this mounts the
  let comp = <Component />;

  return (
    <Amplitude
      eventProperties={{
        view: comp.type.WrappedComponent
          ? comp.type.WrappedComponent.name
          : comp.type.name,
        route: rest.path,
      }}
    >
      <Route
        {...rest}
        render={props =>
          hasToken() ? (
            <Component {...props} />
          ) : (
            <Redirect
              to={{
                pathname: '/login',
                state: {from: window.location.pathname},
              }}
            />
          )
        }
      />
    </Amplitude>
  );
};

export default withRouter(PrivateRoute);
