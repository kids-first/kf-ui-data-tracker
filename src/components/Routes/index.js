import React from 'react';
import {BrowserRouter as Router, Route, Switch} from 'react-router-dom';
import PrivateRoute from './PrivateRoute';

import {
  LoginView,
  InvestigatorListView,
  FileUploadView,
  FileListView,
} from '../../views';

const Routes = () => (
  <Router>
    <Switch>
      <Route path="/login" component={LoginView} />
      <PrivateRoute exact path="/" component={InvestigatorListView} />
      <PrivateRoute path="/file-upload" component={FileUploadView} />
      <PrivateRoute path="/file-list" component={FileListView} />
    </Switch>
  </Router>
);

export default Routes;
