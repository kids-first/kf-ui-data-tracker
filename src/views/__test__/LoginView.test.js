import React from 'react';
import wait from 'waait';
import {render} from 'react-testing-library';
import {MockedProvider} from 'react-apollo/test-utils';
import {MemoryRouter} from 'react-router-dom';
import Routes from '../../Routes';

jest.mock('auth0-js');

it('redirect user back to login view when there is no auth data', async () => {
  // browser mocks
  const localStorageMock = (function() {
    let store = {};
    return {
      getItem: function(key) {
        return store[key] || null;
      },
      setItem: function(key, value) {
        store[key] = value.toString();
      },
      removeItem: function(key) {
        delete store[key];
      },
      clear: function() {
        store = {};
      },
    };
  })();

  Object.defineProperty(window, 'localStorage', {
    value: localStorageMock,
  });

  const tree = render(
    <MockedProvider>
      <MemoryRouter
        initialEntries={['/study/SD_8WX8QQ06/documents/new-documents']}
      >
        <Routes />
      </MemoryRouter>
    </MockedProvider>,
  );
  await wait(10);

  expect(tree.container).toMatchSnapshot();
});
