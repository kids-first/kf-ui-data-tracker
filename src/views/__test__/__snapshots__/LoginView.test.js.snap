// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redirect user back to login view when there is no auth data 1`] = `
<div>
  <div
    class="ui grid"
  >
    <div
      class="eight wide computer fifteen wide mobile twelve wide tablet column"
    >
      <div
        class="ui placeholder segment"
      >
        <div
          class="ui active transition visible inverted dimmer"
          style="display: flex;"
        >
          <div
            class="content"
          >
            <div
              class="ui huge inverted text loader"
            >
              <h2
                class="ui header mt-6 mb-15"
              >
                Kids First Data Tracker
              </h2>
              Redirecing you to the login page...
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="page"
  />
  <div
    class="ui segment footer"
  >
    <div
      class="ui container text-10"
    >
      <div
        class="ui horizontal list"
        role="list"
      >
        <div
          class="item"
          role="listitem"
        >
          <div
            class="ui basic horizontal label text-10"
          />
        </div>
        <div
          class="item"
          role="listitem"
        >
          UI
           
          <a
            href="https://github.com/kids-first/kf-ui-data-tracker/commit/undefined"
            rel="noopener noreferrer"
            target="_blank"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
